api/envoy/config/common/matcher/v3/matcher.proto

```diff
@@ -16,6 +19,93 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 
 // [#protodoc-title: Unified Matcher API]
 
+// A MatchTree defines a matching tree that can be used to match incoming data, arriving at a match action that indicates what action to take in response to the match result.
+// This structure allows for traversal to happen as data is becoming available, allowing for optimized matching.
+//
+// We talk about an "associated operation" throughout this docs: this is the context dependent operation that is configured with this match tree. For example, this could be a HTTP/TCP filter.
+message MatchTree {
+  // The action that should be taken upon matching.
+  message MatchAction {
+    oneof action {
+      option (validate.required) = true;
+
+      // The associated operation should be skipped.
+      bool skip = 1 [(validate.rules).bool = {const: true}];
+
+      // The specified callback value should be passed to the associated operation.
+      string callback = 2;
+    }
+  }
+
+  message MatchLeaf {
+    message LeafMatcher {
+      oneof matcher {
+        option (validate.required) = true;
+
+        MatchPredicate predicate = 1;
+
+        // [#not-implemented-hide:]
+        core.v3.TypedExtensionConfig typed_config = 2;
+      }
+
+      // The action to perform if the match evaluates to true.
+      MatchAction action = 3 [(validate.rules).message = {required: true}];
+    }
+
+    // A linear set of match conditions that will be evaluated in order.
+    // Each matcher will be attempted in order until one succeeds.
+    // If none of the matchers match, the no_match_action will be performed if specified.
+    repeated LeafMatcher matchers = 1 [(validate.rules).repeated = {min_items: 1}];
+
+    // Action to perform if none of the above matchers match.
+    MatchAction no_match_action = 3;
+  }
+
+  message SublinearMatcher {
+    message MultimapMatcher {
+      // The key to apply this matcher for.
+      string key = 1 [(validate.rules).string = {min_len: 1}];
+
+      // The namespace of this key. This is used to distinguish multiple maps
+      // that could be consulted during matching.
+      string key_namespace = 2 [(validate.rules).string = {min_len: 1}];
+
+      // Mapping from exact value matches to the matching subtree to use when
+      // the match applies.
+      // If a header value matches both a prefix and an exact match, the exact
+      // match path will be taken.
+      map<string, MatchTree> exact_matches = 3;
+
+      // Mapping from prefix matches to the matching subtree to use when the match applies.
+      // If a header value matches both a prefix and an exact match, the exact match path will be taken.
+      // If multiple prefixes match, the longest prefix path will be taken.
```

Comment:
So to make sure I understand this correctly, the key of the map is used to do a prefix (or exact) match against the headers? For example, the following config
```
key_namespace: request_headers
key: foo
prefix_matches:
  abc: ...
``` 
will match the header `foo` with value of prefix `abc` (e.g. `abc`, `abcd`, etc.)

What are the reasons to use a map here? Avoid duplicate?